src/com/google/ProductionInstrumentationReporter.java

```diff
@@ -0,0 +1,464 @@
+ * limitations under the License.
+ */
+
+package com.google;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.reflect.TypeToken;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.io.File;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+
+/**
+ * This class will read a file that contains the instrumentation mapping generated by the
+ * compiler production instrumentation pass, and also a list of other files which are the reports
+ * sent by the instrumented production code. It will then take these inputs and generate a single
+ * JSON which provides a detailed breakdown of each instrumentation point.
+ */
+public class ProductionInstrumentationReporter {
+
+  /**
+   * A implementation of the Base64VLQ CharIterator used for decoding the mappings encoded in the
+   * JSON string.
+   */
+  private static class StringCharIterator implements Base64VLQ.CharIterator {
+
+    final String content;
+    final int length;
+    int current = 0;
+
+    StringCharIterator(String content) {
+      this.content = content;
+      this.length = content.length();
+    }
+
+    @Override
+    public char next() {
+      return content.charAt(current++);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return current < length;
+    }
+  }
+
+  private enum InstrumentationType {
+    FUNCTION,
+    BRANCH,
+    BRANCH_DEFAULT;
+
+    public static InstrumentationType fromString(String type) {
+      switch (type) {
+        case "FUNCTION":
+          return FUNCTION;
+        case "BRANCH":
+          return BRANCH;
+        case "BRANCH_DEFAULT":
+          return BRANCH_DEFAULT;
+        default:
+          throw new IllegalArgumentException(
+              "A valid type was not provided in the instrumentation mapping report: " + type);
+      }
+    }
+
+    public static List<InstrumentationType> convertFromStringList(List<String> typesAsString) {
+      return typesAsString.stream()
+          .map(InstrumentationType::fromString)
+          .collect(Collectors.toList());
+    }
+  }
+
+  private static float roundFloat2Decimals(float num) {
+    return (Math.round(num * 10000) / (float) 100);
+  }
+
+  /**
+   * The class contains the data which is provided by the Production Instrumentation pass in the
+   * closure compiler. It will read the data from the file name created with the
+   * --instrument_mapping_report and is required to follow the format used by Closure Compiler class
+   * VariableMap.
+   */
+  private static class InstrumentationMapping {
```

Comment:
Looking at the class - do you need to store fileNames, functionNames and types? I think you can just have `Map<String, Location>` where `Location` is a class with line, column, file name, function name and type. 

You need fileNames and functionNames only during parsing and they can be local variables.